common/connection.go

```diff
@@ -178,31 +179,31 @@ func NewConnection(ctx context.Context, wsURL string, logger *log.Logger) (*Conn
 	return &c, nil
 }
 
-// close cleanly closes the WebSocket connection.
-// Returns an error if sending the close control frame fails.
 func (c *Connection) close(code int) error {
 	c.logger.Debugf("Connection:close", "code:%d", code)
 
 	var err error
 	c.shutdownOnce.Do(func() {
 		defer func() {
-			_ = c.conn.Close()
-
 			// Stop the main control loop
 			close(c.done)
+			_ = c.conn.Close()
 		}()
 
+		c.closeAllSessions()
+
 		err = c.conn.WriteControl(websocket.CloseMessage,
 			websocket.FormatCloseMessage(code, ""),
-			time.Now().Add(10*time.Second),
+			time.Now().Add(time.Second),
 		)
 
-		c.sessionsMu.Lock()
-		for _, s := range c.sessions {
-			s.close()
-			delete(c.sessions, s.id)
-		}
-		c.sessionsMu.Unlock()
+		// According to the WS RFC[1], we might want to wait for a response
+		// Control frame back from the browser here (possibly for the above
+		// timeout duration), but Chrom{e,ium} never sends one, even when
+		// the browser process exits normally after the Browser.close CDP
+		// command. So we don't bother waiting, since it would just needlessly
+		// delay the k6 iteration.
+		// [1]: https://www.rfc-editor.org/rfc/rfc6455#section-1.4
```

Comment:
How are we sure that Chrom{e,ium} never makes a close handshake (sends a close control frame)?

It's because the Chromium code seems like it does so:
* https://chromium.googlesource.com/chromium/src.git/+/master/net/websockets/websocket_channel_test.cc#1100
* https://chromium.googlesource.com/chromium/src.git/+/master/net/websockets/websocket_channel.cc#333

Btw, is it possible that a race condition may cause keep sending messages to the browser after we send the close frame?

Reply:
From my and @imiric chromium *never* sends close frames when using with CDP. Irregardless of how the closing/shutdown has been initiated.  

The links you ahve provided seem to be related to the websockets functionality used for normal websockets connection initiated by javascript code. 

I have no idea if that code is being used by the CDP connection as well :shrug:.

Reply:
> @mstoykov 
> The links you ahve provided seem to be related to the websockets functionality used for normal websockets connection initiated by javascript code.
>
> I have no idea if that code is being used by the CDP connection as well 🤷.

> @imiric 
> I think you're looking at the web implementation, i.e. how the browser itself handles connections initiated from JS. Notice that they mention JS in one of the comments. This might not be the code used by DevTools, though I'm not familiar enough with either C++ or the Chromium code base to know for sure.

DevTools uses the same websocket functionality, see [this](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/devtools/devtools_http_handler.cc;l=181?q=%22devtools%20listening%22&ss=chromium).

It's OK if we are sure since we checked the WS packet trace, etc. 👍